// generalised-datastore/utils/functions.js
"use strict";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

var charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
var combinations = {
  alnum: charset,
  num: "01234556789",
  alpha: "abcdefghijklmnopqrstuvwxyz"
};

var gen_random_int = function gen_random_int(max_int) {
  var min_int = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
  return min_int + Math.floor(Math.random() * max_int);
};

var get_timestamp_from_id = function get_timestamp_from_id(_id) {
  var split_id = _id && _id.split("~");

  return split_id && split_id[2];
};

var generate_random_string = function generate_random_string(len, combination) {
  var string = "";
  combination = combinations[combination] || combinations["num"];

  for (var i = 0; i < (len || 6); i++) {
    string += combination[gen_random_int(combination.length)];
  }

  return string;
};

var _id = function _id(folder) {
  var random_value = "";

  for (var i = 0; i < gen_random_int(32, 12); i++) {
    random_value += charset[gen_random_int(charset.length)];
  }

  return "".concat(folder, "~").concat(random_value, "~").concat(Date.now());
};

var valid_id = function valid_id(val) {
  if (typeof val !== "string" || !val) return;
  var splited = val.split("~");
  if (splited.length !== 3) return;
  return true;
};

var copy_object = function copy_object(object) {
  if (_typeof(object) !== "object") return object;

  var new_object = _objectSpread({}, object);

  for (var prop in new_object) {
    var val = new_object[prop];
    if (Array.isArray(val) && _typeof(val[0]) === "object" && val[0] && valid_id(val[0]._id)) new_object[prop] = val.map(function (v) {
      return copy_object(v);
    });else if (_typeof(val) === "object" && val && !Array.isArray(val)) new_object[prop] = copy_object(val);
  }

  return new_object;
};

module.exports = {
  _id: _id,
  gen_random_int: gen_random_int,
  get_timestamp_from_id: get_timestamp_from_id,
  copy_object: copy_object,
  valid_id: valid_id,
  generate_random_string: generate_random_string
};
